### ISSUE TITLE
SSL peers aren't verfied

### ISSUE THREAD

**Issue by Azolo** [Azolo/websockex#1]
The Erlang `:ssl` module is very strict in its certification checking. 

The result of that is that, if I understand correctly, `verify_peer` option works when everything is basically done correctly.

What I found out is that the internet is obviously broken everywhere, the `:ssl` module reliably connects to just about zero of the sites I tried when verifying the peer with CA certs from [`erlang-certifi`](https://github.com/certifi/erlang-certifi).

So I tried using the `verify_fun` option with [`ssl_verify_fun.erl`](https://github.com/deadtrickster/ssl_verify_fun.erl), plus the `partial_chain` based off of [`hackney`'s implementation][1]... But it didn't work.

After that I tried a bunch of other stuff and made some progress, changed some more stuff only to find that it still didn't work. So I threw my hands up and made everything insecure by default.

### **TLDR;**
I'm not smart enough to make the `:ssl` module's `verify_peer` work correctly.

[1]: https://github.com/benoitc/hackney/blob/ea2012f0d4d97c1ec1a50e840168dff5b48ab97d/src/hackney_connect.erl#L306-L328

**Comment by dmorneau** [comment]
Hackney does it with a couple libraries, certifi and ssl_verify_fun. Perhaps the same would work here?

https://github.com/benoitc/hackney/blob/master/src/hackney_connect.erl#L283

**Comment by xlphs** [comment]
Not sure what you are trying to accomplish but you could simply pass ssl options on. I wrote a [short article](https://mw.gl/posts/elixir_ssl/) on `verify_fun` that shows how to do certificate pinning, hopefully it helps.

**Comment by Azolo** [comment]
@xlphs Cool, thanks for that article. It actually shows me a different way that the `ssl` module could be used and I need to keep that in mind.

What I am trying to accomplish is just being able to open a tcp socket to a site like `https://google.com` and verify that I am in fact connecting to `https://google.com` through the certificate verification. Erlang's `:ssl` module seems super picky about how you do that.

**Comment by khodzha** [comment]
maybe you should take a look at elixir-socket?
i was trying to connect to wss://eu-central434.discord.gg:80, elixir-socket could connect, yet websockex returns %WebSockex.ConnError{original: {:tls_alert, 'record overflow'}}

**Comment by Azolo** [comment]
@khodzha Thanks for the info. I think it's possible that `elixir-socket` is ignoring the `:80` and using port 443 because of the `wss` protocol. I don't think that Discord is handle `SSL` connections on port 80.

**Comment by khodzha** [comment]
yeah, sorry for trouble
discord docs were quite confusing there, connecting without port and with some query params works as expected
:ok_hand:

**Comment by brian-petersen** [comment]
I spent a good part of yesterday working with Erlang's `:ssl` to get it to work with `[verify: :verify_peer]`. I too found it a bit confusing and hard to work with. I followed Hackney's `:ssl` setup as best I could (as documented [here](https://github.com/benoitc/hackney/blob/c8f3a6c5a837e4554be219f7c21ae8fdb3d01c40/src/hackney_ssl.erl#L40)).

I got it working to a point -- but I'm not leveraging https://github.com/deadtrickster/ssl_verify_fun.erl like the Hackney project is. I'm not sure what state this puts it in -- but this is the working code I have so far:

```elixir
def start_link() do
  ssl_options = [
    verify: :verify_peer,
    depth: 99,
    cacertfile: CAStore.file_path(),
    customize_hostname_check: [match_fun: :public_key.pkix_verify_hostname_match_fun(:https)]
  ]

  WebSockex.start_link("wss://example.com/ws", __MODULE__, state, ssl_options: ssl_options) do
end
```

The `CAStore.file_path()` call comes from https://github.com/elixir-mint/castore.

As far I can tell, the only thing really missing from this setup is something like this:

```elixir
ssl_options = [
  verify: :verify_peer,
  depth: 99,
  cacertfile: CAStore.file_path(),
  customize_hostname_check: [match_fun: :public_key.pkix_verify_hostname_match_fun(:https)],
  verify_fun: {&:ssl_verify_hostname.verify_fun/3, [check_hostname: 'URL HERE']},
]
```

(Notice that `check_hostname` is a charlist.) Hope this helps someone!
